File: java-to-go-projects/13-event-emitter/Listener.java
package events;

import java.util.function.Consumer;

/**
 * A listener wraps a callback function and optional metadata.
 * Can be one-shot (auto-removed after first invocation) or persistent.
 */
public class Listener<T> {

    private final String id;
    private final Consumer<T> callback;
    private final boolean once;
    private int invokeCount = 0;

    public Listener(String id, Consumer<T> callback, boolean once) {
        this.id = id;
        this.callback = callback;
        this.once = once;
    }

    /**
     * Invoke the callback with the given data. Returns true if this listener
     * should be removed (one-shot listener that has been called).
     */
    public boolean invoke(T data) {
        callback.accept(data);
        invokeCount++;
        return once;
    }

    public String getId() { return id; }
    public boolean isOnce() { return once; }
    public int getInvokeCount() { return invokeCount; }

    @Override
    public String toString() {
        return String.format("Listener{id='%s', once=%s, invoked=%d}", id, once, invokeCount);
    }
}

File: java-to-go-projects/13-event-emitter/rust/src/listener.rs
use std::sync::atomic::{AtomicUsize, Ordering};

pub struct Listener<T> {
    pub id: usize,
    pub callback: Box<dyn Fn(&T) + Send + Sync>,
    pub once: bool,
    pub invoke_count: AtomicUsize,
}

impl<T> Listener<T> {
    pub fn new(id: usize, callback: Box<dyn Fn(&T) + Send + Sync>, once: bool) -> Self {
        todo!()
    }

    /// Invoke the callback. Returns true if this was a one-shot listener
    /// (should be removed after invocation).
    pub fn invoke(&self, data: &T) -> bool {
        todo!()
    }

    pub fn id(&self) -> usize {
        self.id
    }

    pub fn is_once(&self) -> bool {
        self.once
    }
}
